barretenberg/cpp/src/barretenberg/ecc/groups/element_impl.hpp

```diff
@@ -599,96 +600,91 @@ element<Fq, Fr, T> element<Fq, Fr, T>::mul_without_endomorphism(const Fr& expone
     const uint256_t converted_scalar(exponent);
 
     if (converted_scalar == 0) {
-        element result{ Fq::zero(), Fq::zero(), Fq::zero() };
-        result.self_set_infinity();
-        return result;
+        return element::infinity();
     }
 
-    element work_element(*this);
+    element accumulator(*this);
     const uint64_t maximum_set_bit = converted_scalar.get_msb();
     // This is simpler and doublings of infinity should be fast. We should think if we want to defend against the
     // timing leak here (if used with ECDSA it can sometimes lead to private key compromise)
     for (uint64_t i = maximum_set_bit - 1; i < maximum_set_bit; --i) {
-        work_element.self_dbl();
+        accumulator.self_dbl();
         if (converted_scalar.get_bit(i)) {
-            work_element += *this;
+            accumulator += *this;
         }
     }
-    return work_element;
+    return accumulator;
 }
 
+namespace detail {
+using EndoScalars = std::pair<std::array<uint64_t, 2>, std::array<uint64_t, 2>>;
+template <typename Element, std::size_t NUM_ROUNDS> struct EndomorphismWnaf {
+
+    static constexpr size_t NUM_WNAF_BITS = 4;
+
+    std::array<uint64_t, NUM_ROUNDS * 2> table;
+    bool skew = false;
+    bool endo_skew = false;
+
+    EndomorphismWnaf(const EndoScalars& scalars)
+    {
+        wnaf::fixed_wnaf(&scalars.first[0], &table[0], skew, 0, 2, NUM_WNAF_BITS);
+        wnaf::fixed_wnaf(&scalars.second[0], &table[1], endo_skew, 0, 2, NUM_WNAF_BITS);
+    }
+};
+} // namespace detail
+
 template <class Fq, class Fr, class T>
 element<Fq, Fr, T> element<Fq, Fr, T>::mul_with_endomorphism(const Fr& exponent) const noexcept
 {
+    constexpr size_t NUM_ROUNDS = 32;
     const Fr converted_scalar = exponent.from_montgomery_form();
 
-    if (converted_scalar.is_zero()) {
-        element result{ Fq::zero(), Fq::zero(), Fq::zero() };
-        result.self_set_infinity();
-        return result;
+    if (converted_scalar.is_zero() || is_point_at_infinity()) {
```

Comment:
I'd move it about line 640, then you avoid useless multiplication in conversion from montgomery

Reply:
Sure. I was wondering how much we cared about optimizing infinity points vs readability, but I can move it up